container: support user defined logs endpoint

maxux opened the pull request:
This PR allows user to forward container stdout/stderr to external endpoint. Right now, only redis is supported. You can define on container reservation a new field `logs` which contains information where to push logs:
```
        "logs": [
          {
            "type": "redis",
            "data": {
              "endpoint": "redis://10.241.0.232:6370",
              "channel": "debug"
            }
          }
        ],
```

This endpoint is a reachable redis, and channel is a pub/sub where logs will be sent (one line at a time).

maxux committed:
Add new logger package

maxux committed:
Support logs inside container reservation

maxux committed:
Merge branch 'master' into container-logs

maxux committed:
Apply user defined logs container

maxux committed:
Update mock to support container logs

maxux committed:
Add debug logs target

maxux committed:
Fix logs target exchange between privisiond and contd

maxux committed:
Support container logs from reservation schema

maxux committed:
Add missing comments about structs and functions

maxux committed:
Improve functions comments

maxux committed:
Fix gedis test with container logs

maxux committed:
Fix provision tests with containers logs

maxux committed:
Update provision doc

maxux committed:
Update tfuser to support logs arguments

maxux committed:
Update tfuser changes for tests

maxux committed:
Merge branch 'master' into container-logs

maxux committed:
Merge branch 'master' into container-logs

muhamadazmy commented on pkg/logger/logger_redis.go:
```diff
@@ -0,0 +1,65 @@
+
+// Stderr handle a stderr single line
+func (c *ContainerLoggerRedis) Stderr(line string) error {
+	_, err := c.conn.Do("PUBLISH", c.ChannelStderr, line)
```
Shouldn't we have a different channel for stderr or at least prefix the line with "E:" the same way u do with console logger?

maxux replied:
We support a different channel, it's just not exposed publicly on the reservation.

maxux replied:
Now you can specify different channel for `stdout` and `stderr`

muhamadazmy commented on tools/bcdb_mock/models/generated/workloads/workloads.go:
```diff
@@ -93,9 +93,20 @@ type TfgridWorkloadsReservationContainer1 struct {
 	FarmerTid         int64                                          `bson:"farmer_tid" json:"farmer_tid"`
 }
 
+type TfgridWorkloadsReservationLogs1 struct {
```
Can we add the corresponding schema structures to the files under `models/schema/workloads`. No regeneration is needed, just to be able to port this back to jumpscale.

maxux replied:
Please review my changes to see if models seems right :)

muhamadazmy reviewed (changes requested):
Excellent PR. but I have one comment that requires a change (regarding schema)

zaibon commented on pkg/logger/logger.go:
```diff
@@ -0,0 +1,28 @@
+	Stderr(line string) error
+
+	// Close stdout handler
+	CloseStdout()
```
Is there any reason you want to have a separate method to close stdout and stderr ? We will always want to close both stream no ?

zaibon replied:
I would also return an error here. So we can at least log the error if any.

maxux replied:
Since it's 2 different handler for reading them, I keep both for closing. I guess they will always close at the same time, but nothing guarantee it. And for redis, internally we support two two channels for stdout/stderr, we could also support two files for stdout/stderr, which are not the same, etc. I guess it's better to keep separation ?

maxux replied:
Handled by io.Writer now

zaibon commented on pkg/container/container.go:
```diff
@@ -173,7 +173,39 @@ func (c *containerModule) Run(ns string, data pkg.Container) (id pkg.ContainerID
+
+	// set user defined endpoint logging
+	for _, l := range data.Logs {
+		if l.Type == "redis" {
```
make redis a constant in the logger package

zaibon replied:
you probably want to use a switch here, so where there is new type of logger we can just add a case

zaibon commented on pkg/logger/logger_console.go:
```diff
@@ -0,0 +1,41 @@
+// ContainerLoggerConsole does nothing else that print
+// logs on console stdout/stderr, there are no config
+type ContainerLoggerConsole struct {
+	ContainerLogger
```
you do not need to embed the interface since you type already implements the required method to satisfy the interface

maxux replied:
I can have an empty struct ?

zaibon commented on pkg/logger/logger_file.go:
```diff
@@ -0,0 +1,60 @@
+
+	f, err := os.Create(filepath)
+	if err != nil {
+		return &ContainerLoggerFile{}, err
```
```suggestion
		return nil, err
```

zaibon commented on pkg/logger/logger_file.go:
```diff
@@ -0,0 +1,60 @@
+
+// Stdout handle a stdout single line
+func (c *ContainerLoggerFile) Stdout(line string) error {
+	_, err := c.fd.WriteString(line + "\n")
```
```suggestion
	_, err := fmt.Fprintf(c.fd, "%s\n", line)
```

zaibon commented on pkg/logger/logger_file.go:
```diff
@@ -0,0 +1,60 @@
+
+// Stderr handle a stderr single line
+func (c *ContainerLoggerFile) Stderr(line string) error {
+	_, err := c.fd.WriteString(line + "\n")
```
```suggestion
      _, err := fmt.Fprintf(c.fd, "%s\n", line)
```

zaibon commented on pkg/logger/logger_redis.go:
```diff
@@ -0,0 +1,65 @@
+// ContainerLoggerRedis send stdout/stderr to a
+// Redis PubSub channel
+type ContainerLoggerRedis struct {
+	ContainerLogger
```
no need to embed the interface

zaibon commented on pkg/logger/logger_redis.go:
```diff
@@ -0,0 +1,65 @@
+
+	Endpoint      string
+	ChannelStdout string
+	ChannelStderr string
```
These property doesn't have to be public
```suggestion
	endpoint      string 
  	channelStdout string
	channelStderr string
```

zaibon commented on pkg/logger/logger_file.go:
```diff
@@ -0,0 +1,60 @@
+type ContainerLoggerFile struct {
+	ContainerLogger
+
+	Filepath string
```
```suggestion
	filepath string
```

zaibon commented on pkg/logger/loggers.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+// Wait waits for logs to be closed to clean handlers
+func (c *ContainerLoggers) Wait() {
```
you do not need this method at all. You can just copy the content if this method into the literal function line 96

zaibon commented on tools/tfuser/main.go:
```diff
@@ -244,6 +244,15 @@ func main() {
+							Name:  "logs",
+							Usage: "custom redis endpoint to send process logs (eg: redis://example.tld:1234)",
+						},
+						cli.StringFlag{
```
I guess we could get rid of this flag if we use the `path` of the redis URI. Something like
`redis://host:port/channel`

maxux replied:
I don't know if channel can be specified like that. I'll check log. I maybe need to parse uri and connect not using internal uri connector.

maxux replied:
It's implemented like that now :)

zaibon reviewed (changes requested):
the PR overall looks ok. 
I think there is a way to make things a bit simpler though.
Instead of adding a new ContainerLogger interface, we could just use the [WithStream](https://github.com/containerd/containerd/blob/master/cio/io.go#L117) method. 
This methods takes any `io.Writer`. This means any new logger implementation should just implement the `io.Writer` interface. 

The result would be something like
```go
mw := io.MultiWriter(
	logFile,
	redisLogger,
       // any other object that implement io.Writer...
)
container.NewTask(ctx,cio.NewCreator(cio.WithStreams(nil, mw, mw)))
```

maxux committed:
Apply some review changes

maxux committed:
Merge branch 'master' into container-logs

maxux committed:
Change log implementation for MultiWriter

Replace logs implementation, switching from custom internal handling to
support WithStreams that containerd supports out-of-box. New
implementation leverage this to io.Write class and remove the internal
write process.

maxux commented:
Latest implementation use the `MultiWriter` that @zaibon suggested, reducing amount of code and complexity. Thanks :)

Still some improvement to fix and everything will be updated.

maxux committed:
Rewriting how container logger works and reservation

- Move pkg/logger to pkg/container/logger since logger package are only
for container purpose

- Update name of struct inside logger package for shorter names
  - ContainerLoggerConsole -> Console
  - ContainerLoggerFile    -> File
  - ContainerLoggerRedis   -> Redis

  - NewContainerLoggers       -> NewLoggers
  - NewContainerLoggerFile    -> NewFile
  - NewContainerLoggerRedis   -> NewRedis
  - NewContainerLoggerConsole -> NewConsole

- Update provision to replace endpoint/channel with redis url for stdout
and stderr

- Update tfuser to support new provision fields
  - Removing --logs and --channel
  - Adding --stdout and --stderr options

- Argument are now always using redis://host:port/channel format

maxux committed:
Removing remaing debug fmt.Println

maxux committed:
Add jsx schema in bcdb_mock about logs

maxux committed:
Fix logger lint

maxux committed:
Update code to follow lint

maxux commented:
From 0d6d56f commit message:

- Move `pkg/logger` to `pkg/container/logger` since logger package are only for container purpose
- Update name of struct inside logger package for shorter names
  - `ContainerLoggerConsole` -> `Console`
  - `ContainerLoggerFile` -> `File`
  - `ContainerLoggerRedis` -> `Redis`
  - `NewContainerLoggers` -> `NewLoggers`
  - `NewContainerLoggerFile` -> `NewFile`
  - `NewContainerLoggerRedis` -> `NewRedis`
  - `NewContainerLoggerConsole` -> `NewConsole`
- Update provision to replace endpoint/channel with redis url for stdout and stderr
- Update tfuser to support new provision fields
  - Removing `--logs` and `--channel`
  - Adding `--stdout` and `--stderr` options
- Argument are now always using `redis://host:port/channel` format
